### PR TITLE
doc(compiler): ivy seperate compilation design document

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Documentation content changes
```

Design document for how the ivy compiler will handle metadata for separate compilation.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler/design/separate_compilation.md [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+| `type`              | implicit              |
+| `isComponent`       | `ngComponentDef`      |
+| `selector`          | `ngModuleScope`       |
+| `exportAs`          | `ngDirectiveDef`      |
+| `inputs`            | `ngDirectiveDef`      |
+| `outputs`           | `ngDirectiveDef`      |
+| `hostListeners`     | `ngDirectiveDef`      |
+| `hostProperties`    | `ngDirectiveDef`      |
+| `hostAttributes`    | `ngDirectiveDef`      |
+| `providers`         | `ngInjectableDef`     |
```
Comment (mhevery) [inline packages/compiler/design/separate_compilation.md]: `providers` should be `ngInjectorDef`?
`viewProviders` should be `ngComponentDef`?

**Review by mhevery — APPROVED** [review]

File: packages/compiler/design/separate_compilation.md (hunk 1 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: Yes. Will fix.

File: packages/compiler/design/separate_compilation.md [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+| field               | ivy                   |
+|---------------------|-----------------------|
+| `type`              | implicit              |
+| `name`              | `ngModuleScope`       |
+| `pure`              | `ngPipeDef`           |
+
+The only pieces of information that are not generated into the definition are
+the directive selector and the pipe name as they go into the module scope.
+
+The information needed to build a `ngModuleScope` needs to be communicated
```
Comment (trotyl) [inline packages/compiler/design/separate_compilation.md]: `a` -> `an`?

File: packages/compiler/design/separate_compilation.md [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+from the directive and pipe to the module that declares them.
+
+## Metadata
+
+### Angular 5
+
+Angular 5 uses `.metadata.json` files to store information that is directly
+inferred form the `.ts` files and include value information that is not
```
Comment (trotyl) [inline packages/compiler/design/separate_compilation.md]: `form` -> `from`?

File: packages/compiler/design/separate_compilation.md [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+included in the `.d.ts` file produced by TypeScript. Because only exports for
+types are included in `.d.ts` files and might not include the exports necessary
+for values, the metadata includes export clauses from the `.ts` file.
+
+When a module is flattened into a FESM (Flat EcmaScript Module), a flat
```
Comment (trotyl) [inline packages/compiler/design/separate_compilation.md]: `ECMAScript` as per https://github.com/angular/angular/pull/22316#discussion_r169413900?

File: packages/compiler/design/separate_compilation.md [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+2. Add  `"ngComponentDef": {}` static field.
+3. Add `"ngSelector": <selector-value>` static field.
+
+##### Example
+
+*my.component.ts*
+```ts
+@Component({
+  selector: 'my-comp',
+  template: `<h1>Hello, {{name}}!`
```
Comment (trotyl) [inline packages/compiler/design/separate_compilation.md]: missing `</h1>`?

File: packages/compiler/design/separate_compilation.md [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+component factories are generated as a final global step when compiling
+a module. If any of the transitive information changed then all factories 
+need to be regenerated.
+
+Separate component and module compilation is supported only at the module
+definition level and only from source. That is, npm packages must contain
+the metadata necessary to generate the factories, they cannot contain, 
+themselves, the generated factories. This is because, if any of there
+dependencies change, their factories would be invalid preventing them from
+version ranges in their dependencies. To support producing factories from
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: ranges or changes ?

File: packages/compiler/design/separate_compilation.md [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+compiled source (already translated by TypeScript into JavaScript)
+libraries include metadata that describe the content of the the Angular
+decorators.
+
+### Angular Ivy
+
+In Ivy, the runtime is crafted in a way that allows for separate compilation
+by preforming at runtime much of what was previously pre-calculated by
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: preforming -> performing ?

File: packages/compiler/design/separate_compilation.md [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+compiler. This allows the definition of components to change without
+requiring modules and components that depend on them being recompiled.
+
+The mental model of Ivy is that the decorator is the compiler. That is
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "the decorator is the compiler" I am not a fan of this terminology. Doesn't it bring some confusion ?

File: packages/compiler/design/separate_compilation.md [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+})
+export class MyComponent {
+  @Input() name: string;
+}
+```
+
+*my.component.ts*
+```js
+export class MyComponent {
+  name: string;
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: `static ngSelector` ?

File: packages/compiler/design/separate_compilation.md [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+}
+```
+
+*my.directive.js*
+```js
+export class MyDirective {
+  constructor() {
+    this.dirId = 'some id';
+  }
+  static ngDirectiveDef = defineDirective({...});
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: +`ngSelector`

File: packages/compiler/design/separate_compilation.md [hunk 11]
```diff
@@ -0,0 +1,459 @@
+# DESIGN DOC (Ivy): Separate Compilation
+
+AUTHOR: chuckj@
+
+## Background
+
+### Angular 5
+
+In 5.0 and prior versions of Angular the compiler performs whole program
+analysis and generates template and injector definitions that are use
```
Comment (skreborn) [inline packages/compiler/design/separate_compilation.md]: I believe this was meant to read "injector definitions that are **using** this global knowledge".

File: packages/compiler/design/separate_compilation.md [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+directives into the component, pre-calculate queries, pre-calculate
+content projection, etc. This global knowledge requires that module and
+component factories are generated as a final global step when compiling
+a module. If any of the transitive information changed then all factories 
+need to be regenerated.
+
+Separate component and module compilation is supported only at the module
+definition level and only from source. That is, npm packages must contain
+the metadata necessary to generate the factories, they cannot contain, 
+themselves, the generated factories. This is because, if any of there
```
Comment (skreborn) [inline packages/compiler/design/separate_compilation.md]: Once again, "if any of **their** dependencies change".

File: packages/compiler/design/separate_compilation.md [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+  }
+}
+```
+
+#### `@NgModule`
+
+The metadata for a module is transformed by:
+
+1. Remove the `@NgModule` directive.
+2. Add  `"ngPipeDef": {}` static field.
```
Comment (skreborn) [inline packages/compiler/design/separate_compilation.md]: Did you mean `ngInjectorDef`?

File: packages/compiler/design/separate_compilation.md (hunk 6 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: ranges. You would be required to specify exact versions as you could not tolerate even a point release change.

File: packages/compiler/design/separate_compilation.md (hunk 8 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: It is the model we want to eventually have. That is, we want the decorator to be a class transformer. A transformer is a compiler.

File: packages/compiler/design/separate_compilation.md (hunk 9 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: No. We don't emit the `ngSelector`, it is only present in the metadata. Or, you can think of it as, we pre-remove it. "Manual considerations" below describes how to get build optimizer to remove it for you if you define the component manually.

File: packages/compiler/design/separate_compilation.md (hunk 5 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: Ouch. The compiler in my head is broken.

File: packages/compiler/design/separate_compilation.md (hunk 10 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: See above.

File: packages/compiler/design/separate_compilation.md (hunk 13 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: Cut/paste error.

File: packages/compiler/design/separate_compilation.md (hunk 8 above)
Comment (vicb) [reply packages/compiler/design/separate_compilation.md]: > It is the model we want to eventually have. That is, we want the decorator to be a class transformer. A transformer is a compiler.

I understand the concept. However IMO saying "the decorator is the compiler" without further explanation is confusing. I also find "class transformer" confusing here wrt to TS transformer.

I am not saying that anything is wrong only that it deserves some more explanation.

File: packages/compiler/design/separate_compilation.md [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+export class MyComponent {
+  @Input() name: string;
+}
+```
+
+*my.component.js*
+```js
+export class MyComponent {
+  name: string;
+  static ngComponentDef = defineComponent({...});
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: wrt to my comment about `ngSelector` I understand what you mean. But how does it fit with l138.
Needs more explanation ?

File: packages/compiler/design/separate_compilation.md [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+  transform(...) ...
+}
+```
+
+*my.pipe.js*
+```js
+export class MyPipe {
+  transform(...) ...
+  static ngPipeDef = definePipe({...});
+  static ngSelector = 'myPipe';
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: Why do we have `ngSelector` here ? (cf my comment about more explanation)

File: packages/compiler/design/separate_compilation.md [hunk 16]
```diff
@@ -0,0 +1,790 @@
...
+### Angular Ivy
+
+In Ivy, the runtime is crafted in a way that allows for separate compilation
+by performing at runtime much of what was previously pre-calculated by
+compiler. This allows the definition of components to change without
+requiring modules and components that depend on them being recompiled.
+
+The mental model of Ivy is that the decorator is the compiler. That is
+the decorator can be thought of as parameters to a class transformer that
+transforms the information in the decorator into the corresponding
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "transforms" here is confusing wrt "class transformer".
May be rephrase to say that it transforms the class by generating definitions based on the decorator parameters ?

File: packages/compiler/design/separate_compilation.md [hunk 17]
```diff
@@ -0,0 +1,790 @@
...
+definition. An `@Component` decorator transforms the class by adding
+an `ngComponentDef` static property, `@Directive` adds `ngDirectiveDef`,
+`@Pipe` adds `ngPipeDef`, etc. In most cases values supplied to the
+decorator is sufficient to generate the definition. However, in the case of
+interpreting the template, the compiler needs to know the selector defined for
+each component, directive and pipe that are in scope of the template. The
+purpose of this document is to define what information is needed by the
+compiler that is not provided by the decorator is serialized, discovered and
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: still does not understand this sentence ? "by the decorator is serialized"

File: packages/compiler/design/separate_compilation.md [hunk 18]
```diff
@@ -0,0 +1,790 @@
...
+
+The information available across compilations in Angular 5 is represented in
+the  compiler by a summary description. For example, components and directives
+are represented by the [`CompileDirectiveSummary`](https://github.com/angular/angular/blob/d3827a0017fd5ff5ac0f6de8a19692ce47bf91b4/packages/compiler/src/compile_metadata.ts#L257).
+The following table shows where this information ends up in an ivy compiled
+class:
+
+### `CompileDirectiveSummary`
+
+| field               | ivy                   |
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: find better description for those headers ? (field and ivy)

File: packages/compiler/design/separate_compilation.md [hunk 19]
```diff
@@ -0,0 +1,790 @@
...
+With this proposal, the compiler treats manually and mechanically generated
+Angular definitions identically. This allows flexibility not only in the future
+for how the declarations are mechanically produced it also allows alternative
+mechanism to generate declarations be easily explored without altering the
+compiler or dependent tool chain. It also allows third-party code generators
+with possibly different component syntaxes to generate a component fully
+understood by the compiler.
+
+Unfortunately, however, manually generated module contain references to
+classes that might not be necessary at runtime. Manually or third-party
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: > Manually or third-party ...

I would rephrase this

Something like:

"Manually or third-party components can get the same [...] by removing the the `ngSelector` and `ngModuleScope` properties. When using the build optimizer (add link) this can be achieved by ..."

File: packages/compiler/design/separate_compilation.md [hunk 20]
```diff
@@ -0,0 +1,790 @@
+# DESIGN DOC (Ivy): Separate Compilation
+
+AUTHOR: chuckj@
+
+## Background
+
+### Angular 5
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "Angular 5" -> "Render2" ? May be a quick description of what those terms means in this background section would be great.

File: packages/compiler/design/separate_compilation.md [hunk 21]
```diff
@@ -0,0 +1,790 @@
...
+
+## `ngc` output (non-Bazel)
+
+The cases that `ngc` handle are producing an application and producing a
+reusable library used in an application.
+
+### Application output
+
+The output of the ivy compiler only optionally generates the factories
+generated by the Render2 style output of Angular 5.0. In Ivy, the information
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: use "ivy" or "Ivy" consistently

File: packages/compiler/design/separate_compilation.md [hunk 22]
```diff
@@ -0,0 +1,790 @@
...
+that was generated in factories is now generated in Angular a definition that
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "in Angular a definition" typo

File: packages/compiler/design/separate_compilation.md (hunk 14 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: This document only describes the metadata so 138 describing the transforms to the metadata, not the source. The source is only included here for illustration.

File: packages/compiler/design/separate_compilation.md [hunk 23]
```diff
@@ -0,0 +1,790 @@
...
+is generated as a static field on the Angular decorated class.
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "Angular decorated class" ?

File: packages/compiler/design/separate_compilation.md (hunk 15 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: It shouldn't have been. Removed.

File: packages/compiler/design/separate_compilation.md [hunk 24]
```diff
@@ -0,0 +1,790 @@
...
+
+Render2 requires that, when building the final application, all factories for
+all libraries also be generated. In ivy, the definitions are generated when
+the library is compiled.
+
+The ivy compile can adapt Render2 target libraries by generating the factories
+for them and back-patching, at runtime, the static method into the class.
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: static method -> static prop

File: packages/compiler/design/separate_compilation.md [hunk 25]
```diff
@@ -0,0 +1,790 @@
...
+
+#### Back-patching file (`"render2BackPatching"`)
+
+When an application contains Render2 target libraries (libraries compiled with
+Angular 5.0 or non-ivy 6.0), the ivy definitions need to be back-patch onto the
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: "Angular 5.0 or non-ivy 6.0" -> cf my comment in the "Background section" about defining "Render2" vs "Ivy"

File: packages/compiler/design/separate_compilation.md [hunk 26]
```diff
@@ -0,0 +1,790 @@
...
+component, directive, module, pipe, and injectable classes.
+
+If the Angular compiler option `"render2BackPatching"` is enabled, the compiler
+will generate an `angular.back-patch` module generated to root output
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: rephrase "generate an `angular.back-patch` module generated to root output"

File: packages/compiler/design/separate_compilation.md [hunk 27]
```diff
@@ -0,0 +1,790 @@
...
+directory of the project. If `"generateRender2Factories"` is `true` this
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: I don't understand "If `"generateRender2Factories"` ... "

File: packages/compiler/design/separate_compilation.md [hunk 28]
```diff
@@ -0,0 +1,789 @@
...
+
+| option                       | value    |             |
+|------------------------------|----------|-------------|
+| `"enableIvy"`                | `true`   | required    |
+| `"generateRender2Factories"` | `true`   | implied     |
+| `"render2BackPatching"`      | `true`   | implied     |
+| `"generateCodeForLibraries"` | `true`   | default     |
+| `"annotationsAs"`            | `remove` | implied     |
+| `"enableLegacyTemplate"`     | `false`  | default     |
+| `"preserveWhitespaces"`      | `false`  |             |
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: default

File: packages/compiler/design/separate_compilation.md [hunk 29]
```diff
@@ -0,0 +1,789 @@
...
+
+| option                       | value    |             |
+|------------------------------|----------|-------------|
+| `"enableIvy"`                | `true`   | required    |
+| `"generateRender2Factories"` | `false`  |             |
+| `"render2BackPatching"`      | `false`  | default     |
+| `"generateCodeForLibraries"` | `false`  |             |
+| `"annotationsAs"`            | `remove` | implied     |
+| `"enableLegacyTemplate"`     | `false`  | default     |
+| `"preserveWhitespaces"`      | `false`  |             |
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: default

File: packages/compiler/design/separate_compilation.md [hunk 30]
```diff
@@ -0,0 +1,789 @@
...
+
+To produce a Render2 application the options would look like,
+
+```json
+{
+  "compileOptions": {
+    ...
+  },
+  "angularCompilerOptions": {
+    "target": "application"
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: +1 for the JSON example

File: packages/compiler/design/separate_compilation.md [hunk 31]
```diff
@@ -0,0 +1,789 @@
...
+  "angularCompilerOptions": {
+    "target": "library",
+    "enableIvy": true
+  }
+}
+```
+
+##### example - ivy package
+
+Ivy packages are not supported in Angular 6.0 as they are not recommended as
```
Comment (vicb) [inline packages/compiler/design/separate_compilation.md]: typo "as in"

**Comment by vicb** [comment]
Very nice updates !
<sub>Could you please use fixup commits to push the updates to the PR</sub>

File: packages/compiler/design/separate_compilation.md (hunk 19 above)
Comment (chuckjaz) [reply packages/compiler/design/separate_compilation.md]: This would imply that could just not have it which is exactly the opposite think I am trying to convey.

File: packages/compiler/design/separate_compilation.md [hunk 32]
```diff
@@ -456,4 +456,335 @@ export class MyModule {
...
+invocation contains the root application module. Lazy loaded modules should
+also be considered "application" targets.
+
+The purpose of the "library" target is for are all `ngc` invocations that do
+not contain the root application module or a lazy loaded module.
+
+The purpose of the "package" target is to produce a library package that will
+be an entry point for an npm package. Each entry point should be separately
+compiled using a "package" target. Note that this makes explicit the limitation
+that a module from a package cannot be directly lazily loaded. It must be
```
Comment (mhevery) [inline packages/compiler/design/separate_compilation.md]: maybe add `, because theres is nothing which will trigger the generation of the factory files`

File: packages/compiler/design/separate_compilation.md [hunk 33]
```diff
@@ -0,0 +1,793 @@
...
+rules that depend on other `ng_module` rules.
+
+### Angular Ivy
+
+The `ng_module` rule will still use `ngc` to produce the Angular output but,
+when producing ivy output, it no longer will need the `.ngsummary.json` file.
+
+#### `ivy_sources`
+
+The `ivy_sources` can be used as use to cause the ivy versions of files to be
```
Comment (mhevery) [inline packages/compiler/design/separate_compilation.md]: This section is confusing. Could you add an example code, I think that would make it more clear.
